### PR TITLE
#6410 QA corrections for Refresh Emission Views MATSHG and NOXRATECEMS.

### LINKED ISSUE #6410: Emission Views SP Partition Speed Solution
## Background

ECMPS uses Emission View tables to store display ready versions of emissions data.  The tables are needed because formatting the reported data for display takes too much time for a user expecting to see the data.  Because the emissions data exists in both the Workspace and Official schemas two sets of the Emission View tables are needed.  A set of stored procedures were created to populate the Workspace schema Emission View tables and those procedures were performant.  A version of those SP were created to populate the Official schema Emission View tables, but those procedures where not performant.

I believe the initial understanding of the Official schema performance problem was primarily because of the much larger amount of data in the Official schema.  Additionally and based on testing, the multiple joins on the single emission tables for different parameters appeared to exacerbate slowness.  Because of the perceived join issue, the "pivot" method was developed, which used table functions to flatten the hourly rows for multiple parameters into single rows for each hour to require one join per table.  Although the "pivot" method reduced the slowness issue, speed problems with some of the Official stored procedures still exist.

Additional research exposed some truths about querying the RPT_PERIOD_ID partitioned Official schema emission tables.  Where and especially join conditions on the tables will result in searching each partition for a table if the RPT_PERIOD_ID is not included in the condition for the table.  This was the original slowness problem and explains why the "pivot" method worked since it included the RPT_PERIOD_ID in the where clause it used.

### Previous Tickets

This ticket is intended to replace previous tickets and will result in a QA of all Emission Views and will cover the testing for those tickets.  As a result, both of the tickets below will be closed.

#5793
#6324

## Requirement

Update the stored procedures that load the Emission View tables to use direct joins for each parameter to an emissions table.  However each join to the tables **must**  include the RPT_PERIOD_ID.

Additionally, after ensuring they are not used outside of Emission View population, drop the following table functions in both the CAMDECMPS and CAMDECMPSWKS schemas.

- GET_DERIVED_HOURLY_VALUES_PIVOTED( varchar, numeric, text )
- GET_HOURLY_PARAM_FUEL_FLOW_PIVOTED( varchar, numeric, text )
- GET_MONITOR_HOURLY_VALUES_PIVOTED( text, numeric, text )
- GET_MONITOR_HOURLY_VALUES_PIVOTED( varchar, numeric, text, text, text )

## Stored Procedure List

Handled SP already include the adjustment to use the RPT_PERIOD_ID in any where or on clause that references a partitioned emissions table.

| Stored Procedure | Screen Name | Official | Workspace |
| :--- | :--- | :--: | :--: |
| REFRESH_EMISSION_VIEW_ALL | Hourly Combined Parameters View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_CO2APPD | Hourly CO2 Appendix D View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_CO2CALC | CO2 Calculation | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_CO2CEMS | Hourly CO2 CEMS View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_CO2DAILYFUEL | CO2 Daily Fuel Sampling View | :red_square: | :red_square: |
| REFRESH_EMISSION_VIEW_DAILYCAL | Daily Calibration View | :ballot_box_with_check: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_HIAPPD | Hourly Heat Input Appendix D View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_HICEMS | Hourly Heat Input CEMS View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_HIUNITSTACK | Heat Input for Unit/Stack View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_LME | LME View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_MASSOILCALC | Mass Oil Calculation View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_MATSHCL | MATS HCL View | :ballot_box_with_check: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_MATSHF | MATS HF View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_MATSHG | MATS HG View | :ballot_box_with_check: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_MATSSO2 | MATS SO2 View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_MATSSORBENT | MATS Sorbent View | :red_square: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_MATSWEEKLY | MATS weekly View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_MOISTURE | Moisture View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_NOXAPPEMIXEDFUEL | Unit Level Fuel Curve View | :red_square: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_NOXAPPESINGLEFUEL | NOX Appendix E Individual Fuel Curve View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_NOXMASSCEMS | Hourly NOx Mass CEMS View | :ballot_box_with_check: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_NOXRATECEMS | Hourly NOx Rate CEMS View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_OTHERDAILY | Other Daily Tests View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_SO2APPD | Hourly SO2 Appendix D View | :white_check_mark: | :white_check_mark: |
| REFRESH_EMISSION_VIEW_SO2CEMS | Hourly SO2 CEMS View | :white_check_mark: | :white_check_mark: |

### Additional Issues
1. **CO2DAILYFUEL** and **NSPS4T** has data in **CAMDECMPS** after test run of "**refresh_emissions_views**" for the following Monitoring Plan and Reporting Period Ids, but data did not exist in **CAMDECMPSWKS** after the test run.
   - ( 'TWCORNEL5-C0E3879920A14159BAA98E03F1980A7A'  , 123 ),
   - ( 'MDC-19939DF8A9FD4C12B3869C394A2C0B1D'        , 123 ),
   - ( 'EGWITLT01-C88ABE75BA5D4C6CB5E2D1EFEC8C4F38'  , 123 ),
   - ( 'MDC-0B5C7E00DEFD410A9B883B5E381A8C4B'        , 113 ),
   - ( 'MDC-80A407D4AC8D478E93F95FB702EC0B3E'        , 123 ),
   - ( 'BPU094-545C837365784BE781EDECEAA08626A3'     , 123 ),
   - ( 'MDC-07AC944F78A043E08E21977F9DA727EA'        , 123 ),
   - ( 'MDC-4F0130B46ADB4B4F9C2EE0BA4A72D463'        , 123 ),
   - ( 'MDC-88B24CFAC6E94EFAA0517A9EC9FB4D9E'        , 123 ),
   - ( 'MDC-98642474BB9D49CAB5CB11F2352A3598'        , 123 ),
   - ( 'RMUENSTERL-FBE23B81FE4A4C6DADBDDAA1496DB309' , 123 ),
   - ( 'MDC-EC0E1E1664394A6DB9DF0DBCFD5C93C1'        , 123 ),
   - ( 'MDC-3B23A43070D74F53AA2DAD83ABFC190C'        , 123 )

### PR DESCRIPTION
**Pull request by djw4erg**
### Notes

1. The CAMDECMPS and CAMDECMPSWKS versions of the SP should be the same except for schema names.
2. The MATSHG SP contains two changes:
   1.  Include the SORBENT_TRAP_SUPP_DATA/SAMPLING_TRAIN_SUPP_DATA to determine Component Ids.
   2. Correct display value for MATS_STARTUP_SHUTDOWN.
3. The NOXRATECEMS SP contains one change to allow using O2 data without a Moisture Basis as either Dry or Wet.

### CODE REVIEW

File: camdecmpswks/procedures/refresh_emission_view_matshg.sql [hunk 1]
```diff
@@ -25,43 +25,80 @@ BEGIN
...
+                                        join camdecmpswks.SORBENT_TRAP trp
+                                          on trp.rpt_period_id = ems.rpt_period_id 
+                                         and trp.mon_loc_id = mpl.mon_loc_id
+                                         and trp.begin_date <= prd.end_date
+                                         and trp.end_date >= prd.begin_date
+                                        join camdecmpswks.SAMPLING_TRAIN trn
+                                          on trn.rpt_period_id = ems.rpt_period_id 
+                                         and trn.trap_id = trp.trap_id
+                                        join camdecmpswks.COMPONENT cmp
+                                          on cmp.component_id = trn.component_id
```
Comment (maxdiebold-erg) [inline camdecmpswks/procedures/refresh_emission_view_matshg.sql]: This join is not in the corresponding CAMDECMPS procedure, should it be? It doesn't look like it's necessary.

File: camdecmpswks/procedures/refresh_emission_view_matshg.sql [hunk 2]
```diff
@@ -25,43 +25,80 @@ BEGIN
...
+                                        join camdecmpswks.MONITOR_PLAN_LOCATION mpl
+                                          on mpl.mon_plan_id= ems.mon_plan_id 
+                                        join camdecmpswks.SORBENT_TRAP_SUPP_DATA trp
+                                          on trp.mon_loc_id = mpl.mon_loc_id
+                                         and trp.begin_date <= prd.end_date
+                                         and trp.end_date >= prd.begin_date
+                                        join camdecmpswks.SAMPLING_TRAIN_SUPP_DATA trn
+                                          on trn.trap_id = trp.trap_id
+                                        join camdecmpswks.COMPONENT cmp
+                                          on cmp.component_id = trn.component_id
```
Comment (maxdiebold-erg) [inline camdecmpswks/procedures/refresh_emission_view_matshg.sql]: This is not in the corresponding CAMDECMPS procedure, should it be? It doesn't look like it's necessary.

**Review by maxdiebold-erg — CHANGES_REQUESTED** [review]
Looks good, I just have a couple question which I've noted below.

Also, could you add the migration scripts to the deployments folder?

File: camdecmpswks/procedures/refresh_emission_view_matshg.sql (hunk 1 above)
Comment (djw4erg) [reply camdecmpswks/procedures/refresh_emission_view_matshg.sql]: Good catch Max!  Removing, will retest, and request another review.

File: camdecmpswks/procedures/refresh_emission_view_matshg.sql (hunk 2 above)
Comment (djw4erg) [reply camdecmpswks/procedures/refresh_emission_view_matshg.sql]: Good catch Max!  Removing, will retest, and request another review.

**Review by maxdiebold-erg — APPROVED** [review]

**Pull request merged**